File: 3561-remove-methods-from-project/remove-methods-from-project.py
class Solution:
    def remainingMethods(self, n: int, k: int, invocations: List[List[int]]) -> List[int]:

        adj = [[] for _ in range(n)]
        indegree=[0]*n
        suspicious = [False]*n
        for u, v in invocations:
            adj[u].append(v)
            indegree[v] += 1

        queue = deque([k])
        suspicious[k] = True

        while queue:
            curr = queue.popleft()
            for neighbor in adj[curr]:
                indegree[neighbor] -= 1

                if not suspicious[neighbor]:
                    suspicious[neighbor] = True
                    queue.append(neighbor)

        result = []
        cannot_remove = False

        for i in range(n):
            if suspicious[i] and indegree[i] > 0:
                cannot_remove = True
                break

            if not suspicious[i]:
                result.append(i)

        if cannot_remove:
            return list(range(n))

        return result

        

File: 3561-remove-methods-from-project/remove-methods-from-project.cpp
class Solution {
public:
    vector<int> remainingMethods(int n, int k, vector<vector<int>>& invocations) {
        vector<vector<int>> adj(n);
        vector<int> indegree(n, 0);
        vector<bool> sus(n, false);

        for(auto &edge: invocations){
            int u = edge[0];
            int v = edge[1];
            adj[u].push_back(v);
            indegree[v]++;
        }

        queue<int> q;
        q.push(k);
        sus[k] = true;

        while(!q.empty()){
            int curr = q.front();
            q.pop();
            for(int &ngbr: adj[curr]){
                indegree[ngbr]--;
                if(!sus[ngbr]){
                    q.push(ngbr);
                    sus[ngbr]=true;
                }
            }
        }

        vector<int> res;
        bool cannotRemove=false;
        for(int i=0;i<n;i++){
            if(sus[i] && indegree[i]>0){
                cannotRemove = true;
                break;
            }
            if(!sus[i]) res.push_back(i);
        }

        if(cannotRemove){
            vector<int> vec(n);
            for(int i=0;i<n;i++){
                vec[i] = i;
            }
            return vec;
        }

        return res;
    }
};